server/src/lsp-server.ts

```diff
@@ -673,19 +674,30 @@ export class LspServer {
         }
     }
 
-    async codeAction(params: lsp.CodeActionParams): Promise<(lsp.Command | lsp.CodeAction)[]> {
+    async codeAction(params: lsp.CodeActionParams): Promise<lsp.CodeAction[]> {
         const file = uriToPath(params.textDocument.uri);
         this.logger.log('codeAction', params, file);
         if (!file) {
             return [];
         }
         const args = toFileRangeRequestArgs(file, params.range);
-        const codeActions: (lsp.Command | lsp.CodeAction)[] = [];
-        const errorCodes = params.context.diagnostics.map(diagnostic => Number(diagnostic.code));
-        provideQuickFix(await this.getCodeFixes({ ...args, errorCodes }), codeActions, this.documents);
-        provideRefactors(await this.getRefactors(args), codeActions, args);
-        provideOrganizeImports(file, params.context, codeActions);
-        return codeActions;
+        const actions: lsp.CodeAction[] = [];
+        if (!params.context.only || params.context.only.findIndex((k: string) => k.startsWith(CodeActionKind.QuickFix)) > -1) {
+            const errorCodes = params.context.diagnostics.map(diagnostic => Number(diagnostic.code));
+            actions.push(...provideQuickFix(await this.getCodeFixes({ ...args, errorCodes }), this.documents));
+        }
+        if (!params.context.only || params.context.only.findIndex((k: string) => k.startsWith(CodeActionKind.Refactor)) > -1) {
+            actions.push(...provideRefactors(await this.getRefactors(args), args));
+        }
+
+        // organize import is provided by tsserver for any line, so we only get it if explicitly requested
+        if (params.context.only && params.context.only.indexOf(CodeActionKind.SourceOrganizeImports) > -1) {
```

Comment:
The logic for checking `only` should be the same here as in other cases - should use `startsWith`. Otherwise, the code action won't be provided if the client asks for `only: ['source']`.